Ambiguous Dependencies with PMD

I am working on moving an existing internal Jenkins plugin from using the JPI plugin to using the JPI2 plugin and have run into an issue.

When combining the JPI2 plugin with PMD, Gradle reports ambiguous dependencies for the PMD configurations. It appears that the JPI2 plugin is injecting the Jenkins plugin dependencies to the classpath, but not helping Gradle disambiguate which ones it needs.

<details>
<summary>Check Output</summary>

```
$ ./gradlew check
> Task :pmdTest FAILED

[Incubating] Problems report is available at: file:///mnt/c/Dev/Scratch/jpi-plugin-tests/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':pmdTest' (registered by plugin 'org.gradle.pmd').
> Could not resolve all files for configuration ':testPmdAuxClasspath'.
   > Could not resolve org.jenkins-ci.plugins.workflow:workflow-cps:4331.v9d06ed4658ff.
     Required by:
         root project 'jpi-plugin-tests'
      > The consumer was configured to find a library for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally. There are several available matching variants of org.jenkins-ci.plugins.workflow:workflow-cps:4331.v9d06ed4658ff
        The only attribute distinguishing these variants is 'org.jenkinsci.gradle.plugins.jpi2.artifact.type'. Add this attribute to the consumer's configuration to resolve the ambiguity:
          - Value: 'default' selects variant: 'defaultRuntime'
          - Value: 'pluginJar' selects variant: 'runtime'
   > Could not resolve org.jenkins-ci.plugins.workflow:workflow-durable-task-step:1479.v56e587f413a_7.
     Required by:
         root project 'jpi-plugin-tests'
      > The consumer was configured to find a library for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally. There are several available matching variants of org.jenkins-ci.plugins.workflow:workflow-durable-task-step:1479.v56e587f413a_7
        The only attribute distinguishing these variants is 'org.jenkinsci.gradle.plugins.jpi2.artifact.type'. Add this attribute to the consumer's configuration to resolve the ambiguity:
          - Value: 'default' selects variant: 'defaultRuntime'
          - Value: 'pluginJar' selects variant: 'runtime'
   > Could not resolve org.jenkins-ci.plugins.workflow:workflow-job:1571.1580.v18e46842c125.
     Required by:
         root project 'jpi-plugin-tests'
      > The consumer was configured to find a library for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally. There are several available matching variants of org.jenkins-ci.plugins.workflow:workflow-job:1571.1580.v18e46842c125
        The only attribute distinguishing these variants is 'org.jenkinsci.gradle.plugins.jpi2.artifact.type'. Add this attribute to the consumer's configuration to resolve the ambiguity:
          - Value: 'default' selects variant: 'defaultRuntime'
          - Value: 'pluginJar' selects variant: 'runtime'

* Try:
> Ambiguity errors are explained in more detail at https://docs.gradle.org/9.6.0/userguide/variant_model.html#sub:variant-ambiguity.
> Review the variant matching algorithm at https://docs.gradle.org/9.6.0/userguide/variant_attributes.html#sec:abm_algorithm.
> Use the dependencyInsight report with the --all-variants option to view all variants of the ambiguous dependency.  This report is described at https://docs.gradle.org/9.6.0/userguide/viewing_debugging_dependencies.html#sec:identifying_reason_dependency_selection.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights from a Build Scan (powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.6.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 6s
14 actionable tasks: 2 executed, 12 up-to-date
```
</details>

See [the replication repository](https://github.com/mtughan/jpi-plugin-tests/tree/pmd) for more details.

--- a/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/ArtifactTypeDisambiguationRule.java
+++ b/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/ArtifactTypeDisambiguationRule.java
@@ -1,0 +1,32 @@
+package org.jenkinsci.gradle.plugins.jpi2;
+
+import org.gradle.api.attributes.AttributeDisambiguationRule;
+import org.gradle.api.attributes.MultipleCandidatesDetails;
+import org.jetbrains.annotations.NotNull;
+
+import javax.inject.Inject;
+
+/**
+ * Third-party plugins (e.g. {@code pmd}) resolve their own configurations over hpi/jpi-packaged
+ * dependencies without requesting {@link ArtifactType#ARTIFACT_TYPE_ATTRIBUTE}, which otherwise
+ * leaves {@link HpiMetadataRule}'s "runtime" and "defaultRuntime" variants ambiguous. When no
+ * value is requested, prefer the plain jar variant, matching what jenkinsPlugin-owned
+ * configurations already request explicitly.
+ */
+abstract class ArtifactTypeDisambiguationRule implements AttributeDisambiguationRule<ArtifactType> {
+
+    @Inject
+    public ArtifactTypeDisambiguationRule() {
+    }
+
+    @Override
+    public void execute(@NotNull MultipleCandidatesDetails<ArtifactType> details) {
+        if (details.getConsumerValue() != null) {
+            return;
+        }
+        details.getCandidateValues().stream()
+                .filter(candidate -> ArtifactType.PLUGIN_JAR.equals(candidate.getName()))
+                .findFirst()
+                .ifPresent(details::closestMatch);
+    }
+}

--- a/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/V2JpiPlugin.java
+++ b/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/V2JpiPlugin.java
@@ -245,6 +245,7 @@ public class V2JpiPlugin implements Plugin<Project> {
         dependencies.add(jenkinsCore.getName(), jenkinsCoreCoordinate);
 
         dependencies.getComponents().all(HpiMetadataRule.class);
+        dependencies.getAttributesSchema().attribute(ARTIFACT_TYPE_ATTRIBUTE).getDisambiguationRules().add(ArtifactTypeDisambiguationRule.class);
         var publicationName = configurePublishing(project, jpiTask, defaultRuntime, extension);
 
         var publishing = project.getExtensions().getByType(PublishingExtension.class);

--- a/jpi2/src/test/java/org/jenkinsci/gradle/plugins/jpi2/PmdInteroperabilityIntegrationTest.java
+++ b/jpi2/src/test/java/org/jenkinsci/gradle/plugins/jpi2/PmdInteroperabilityIntegrationTest.java
@@ -1,0 +1,68 @@
+package org.jenkinsci.gradle.plugins.jpi2;
+
+import org.jenkinsci.gradle.plugins.jpi.IntegrationTestHelper;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledOnOs;
+import org.junit.jupiter.api.condition.OS;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DisabledOnOs(value = OS.WINDOWS, disabledReason = "TempDir doesn't appear to work correctly on Windows")
+class PmdInteroperabilityIntegrationTest extends V2IntegrationTestBase {
+
+    /**
+     * HpiMetadataRule adds a "defaultRuntime" variant (attributed {@code ArtifactType.DEFAULT})
+     * alongside the existing "runtime" variant (attributed {@code ArtifactType.PLUGIN_JAR}) to
+     * every hpi/jpi-packaged dependency. Configurations the plugin controls (runtimeClasspath,
+     * testRuntimeClasspath) request {@code ArtifactType.PLUGIN_JAR} explicitly to disambiguate,
+     * but the `pmd` plugin resolves its own aux classpath configuration, which never requests
+     * that attribute. ArtifactTypeDisambiguationRule resolves the ambiguity for such consumers
+     * by defaulting to the plain jar variant when nothing is requested.
+     */
+    @Test
+    void pmdTestResolvesAuxClasspathForJpiPluginDependency() throws IOException {
+        // given
+        var ith = new IntegrationTestHelper(tempDir, "8.14");
+        initBuild(ith);
+        Files.writeString(ith.inProjectDir("build.gradle.kts").toPath(), """
+                plugins {
+                    id("org.jenkins-ci.jpi2")
+                    pmd
+                }
+                repositories {
+                    mavenCentral()
+                    jenkinsPublic()
+                }
+                tasks.withType(Test::class) {
+                    useJUnitPlatform()
+                }
+                group = "com.example"
+                version = "1.0.0"
+                dependencies {
+                    testImplementation("org.jenkins-ci.plugins:git:5.7.0")
+                    testImplementation("org.junit.jupiter:junit-jupiter:5.11.4")
+                    testRuntimeOnly("org.junit.platform:junit-platform-launcher")
+                }
+                """, StandardCharsets.UTF_8);
+        ith.mkDirInProjectDir("src/test/java/com/example");
+        Files.writeString(ith.inProjectDir("src/test/java/com/example/ExampleTest.java").toPath(), """
+                package com.example;
+                import org.junit.jupiter.api.Test;
+                class ExampleTest {
+                    @Test
+                    void example() {
+                    }
+                }
+                """, StandardCharsets.UTF_8);
+
+        // when
+        var result = ith.gradleRunner().withArguments("pmdTest").build();
+
+        // then
+        assertThat(result.getOutput()).contains("BUILD SUCCESSFUL");
+    }
+}

